doc/cascadia/Keybindings-Arguments.md

```diff
@@ -0,0 +1,315 @@
+{ "keys": ["ctrl+alt+shift+1"], "command": "splitProfile", "args": { "orientation":"horizontal", "profileIndex": 0 } },
+{ "keys": ["ctrl+alt+2"], "command": "splitProfile", "args": { "orientation":"vertical", "profileIndex": 1 } },
+{ "keys": ["ctrl+alt+shift+2"], "command": "splitProfile", "args": { "orientation":"horizontal", "profileIndex": 1 } },
+// etc...
+```
+
+Note that instead of having 9 different `newTabProfile<N>` actions, we have a
+singular `newTabProfile` action, and that action requires a `profileIndex` in
+the `args` object.
+
+Also, pay attention to the last set of keybindings, the `splitProfile` ones.
+This is a function that requires two arguments, both a `orientation` and a
+`profileIndex`. Before this change we would have needed to create 20 separate
+actions (10 profile indicies * 2 directions) to handle these cases. Now it can
+be done with a single action that can be much more flexible in its
+implementation.
+
+### Parsing KeyBinding Arguments
+
+We'll add a new `IKeyBindingArgs` interface. All current keybinding events will
+be changed from their current types to `TypedEventHandler`s. These
+`TypedEventHandler`s second param will always be an instance of
+`IKeyBindingArgs`. So for example:
+
+```csharp
+
+delegate void CopyTextEventArgs();
+delegate void NewTabEventArgs();
+delegate void NewTabWithProfileEventArgs(Int32 profileIndex);
+// ...
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event CopyTextEventArgs CopyText;
+    event NewTabEventArgs NewTab;
+    event NewTabWithProfileEventArgs NewTabWithProfile;
+```
+
+Becomes:
+
+```csharp
+interface IKeyBindingArgs {
+    Boolean Handled;
+}
+
+runtimeclass CopyTextEventArgs : IKeyBindingArgs
+{
+    Boolean CopyWhitespace;
+}
+runtimeclass NewTabEventArgs : IKeyBindingArgs { }
+runtimeclass NewTabWithProfileEventArgs : IKeyBindingArgs
+{
+    Int32 ProfileIndex;
+}
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, CopyTextEventArgs> CopyText;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabEventArgs> NewTab;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabWithProfileEventArgs> NewTabWithProfile;
+```
+
+We'll also change our existing map in the `AppKeyBindings` implementation.
+Currently, it's a `std::unordered_map<KeyChord, ShortcutAction, ...>`, which
+uses the `KeyChord` to lookup the `ShortcutAction`. We'll need to introduce a
+new type `ActionAndArgs`:
+
+```csharp
+runtimeclass ActionAndArgs
+{
+    ShortcutAction action;
+    IKeyBindingArgs args;
+}
+```
+
+and we'll change the map in `AppKeyBindings` to a `std::unordered_map<KeyChord,
+ActionAndArgs, ...>`.
+
+When we're parsing keybindings, we'll need to construct args for each of the
+events to go with each binding. When we find some key chord bound to a given

```

Comment:
this is a little light on details about _how_ they'll be parsed. How do we figure out which deserializer to use? How does this accrue to an inevitable extension model where we may not be able to pass a `json::value` across the winrt type boundary?

Reply:
I'll add an example to make it clearer. 

re: extensions: I hate myself for saying this, but maybe a `JsonValue` <-> `windows.data.json.jsonobject` converter? We'll parse the json with jsoncpp, then convert the tree to a winrt `JsonObject`that the extension could read? It's not good but it's possible without forcing ever extension to ship with `jsoncpp` (or using `Windows.Data.Json`) themselves. I could add a note in the Future Considerations on this. At this rate, we'll be able to put together an extensions spec just from the "Future Considerations" of every other spec :P